token/program-2022/tests/transfer.rs

```diff
@@ -0,0 +1,240 @@
+    assert_eq!(
+        error,
+        TokenClientError::Client(Box::new(TransportError::TransactionError(
+            TransactionError::InstructionError(
+                0,
+                InstructionError::Custom(TokenError::OwnerMismatch as u32)
+            )
+        )))
+    );
+}
+
+#[tokio::test]
+async fn self_transfer() {
+    let mut context = TestContext::new().await;
+    context.init_token_with_mint(vec![]).await.unwrap();
+    let TokenContext {
+        decimals,
+        mint_authority,
+        token,
+        alice,
+        ..
+    } = context.token_context.unwrap();
+
+    let alice_account = Keypair::new();
+    let alice_account = token
+        .create_auxiliary_token_account(&alice_account, &alice.pubkey())
+        .await
+        .unwrap();
+
+    // mint a token
+    let amount = 10;
+    token
+        .mint_to(&alice_account, &mint_authority, amount)
+        .await
+        .unwrap();
+
+    // self transfer is ok
+    token
+        .transfer_checked(&alice_account, &alice_account, &alice, 1, decimals)
+        .await
+        .unwrap();
+    token
+        .transfer_unchecked(&alice_account, &alice_account, &alice, 1)
+        .await
+        .unwrap();
+
+    // too much self transfer is not ok
+    let error = token
+        .transfer_checked(&alice_account, &alice_account, &alice, amount + 1, decimals)
+        .await
+        .unwrap_err();
+    assert_eq!(
+        error,
+        TokenClientError::Client(Box::new(TransportError::TransactionError(
+            TransactionError::InstructionError(
+                0,
+                InstructionError::Custom(TokenError::InsufficientFunds as u32)
+            )
+        )))
+    );
+}
+
+#[tokio::test]
+async fn self_owned() {
+    let mut context = TestContext::new().await;
+    context.init_token_with_mint(vec![]).await.unwrap();
+    let TokenContext {
+        decimals,
+        mint_authority,
+        token,
+        alice,
+        bob,
+        ..
+    } = context.token_context.unwrap();
+
+    let alice_account = token
+        .create_auxiliary_token_account(&alice, &alice.pubkey())
+        .await
+        .unwrap();
+    let bob_account = Keypair::new();
+    let bob_account = token
+        .create_auxiliary_token_account(&bob_account, &bob.pubkey())
+        .await
+        .unwrap();
+
+    // mint a token
+    let amount = 10;
+    token
+        .mint_to(&alice_account, &mint_authority, amount)
+        .await
+        .unwrap();
+
+    // unchecked is ok
+    token
+        .transfer_unchecked(&alice_account, &bob_account, &alice, 1)
+        .await
+        .unwrap();
+
+    // checked is ok
+    token
+        .transfer_checked(&alice_account, &bob_account, &alice, 1, decimals)
+        .await
+        .unwrap();
+
+    // self transfer is ok
+    token
+        .transfer_checked(&alice_account, &alice_account, &alice, 1, decimals)
+        .await
+        .unwrap();
+}
+
+#[tokio::test]
+async fn transfer_fee() {
```

Comment:
nit: I haven't hit on the best alternative, but I found this test name confusing, since there is no fee moving.

Reply:
`transfer_with_fee_on_mint_without_fee`?

Reply:
wfm!